zebra-state/src/service.rs

```diff
@@ -718,98 +635,53 @@ impl StateService {
     /// Attempt to validate and commit all queued blocks whose parents have
     /// recently arrived starting from `new_parent`, in breadth-first ordering.
     #[tracing::instrument(level = "debug", skip(self, new_parent))]
-    fn process_queued(&mut self, new_parent: block::Hash) {
-        let mut new_parents: Vec<(block::Hash, Result<(), CloneError>)> =
-            vec![(new_parent, Ok(()))];
+    fn send_process_queued(&mut self, new_parent: block::Hash) {
+        use tokio::sync::mpsc::error::SendError;
+        if let Some(non_finalized_block_write_sender) = &self.non_finalized_block_write_sender {
+            let mut new_parents: Vec<block::Hash> = vec![new_parent];
+
+            while let Some(parent_hash) = new_parents.pop() {
+                let queued_children = self
+                    .queued_non_finalized_blocks
+                    .dequeue_children(parent_hash);
+
+                for queued_child in queued_children {
+                    let child_hash = queued_child.0.hash;
+
+                    let send_result = non_finalized_block_write_sender.send(
+                        NonFinalizedWriteCmd::ProcessQueued {
+                            parent_hash,
+                            queued_child,
+                        },
+                    );
 
-        while let Some((parent_hash, parent_result)) = new_parents.pop() {
-            let queued_children = self
-                .queued_non_finalized_blocks
-                .dequeue_children(parent_hash);
+                    if let Err(SendError(NonFinalizedWriteCmd::ProcessQueued {
+                        parent_hash: _,
+                        queued_child: (_, rsp_tx),
+                    })) = send_result
+                    {
+                        // If Zebra is shutting down, ignore dropped block result receivers
+                        let _ = rsp_tx.send(Err(
+                            "block commit task exited. Is Zebra shutting down?".into(),
+                        ));
```

Comment:
If the channel is closed, we also need to send errors on all the queued blocks, because they will never verify.
See `clear_finalized_block_queue()` for an example.

We also need to call this cleanup method in the `Drop` impl.
(I need to do a similar change in PR #5134.)

Reply:
Good point. Fixed in [`3aad1b7`](https://github.com/ZcashFoundation/zebra/pull/5257/commits/3aad1b7e80f413c12ff67efca8b601d858992036)

Reply:
I tweaked the error messages in 6e2c3b1 in PR #5299.